lms/validation/blackboard.py

```diff
@@ -0,0 +1,26 @@
+"""
+Schemas for Blackboard API responses.
+
+See: https://developer.blackboard.com/portal/displayApi
+"""
+from marshmallow import INCLUDE, Schema, fields, post_load
+
+from lms.validation._base import RequestsResponseSchema
+
+
+class BlackboardListFilesSchema(RequestsResponseSchema):
+    """Schema for Blackboard API /courses/{courseId}/resources responses."""
+
+    class ResultsSchema(Schema):
+        class Meta:
+            unknown = INCLUDE
+
+        id = fields.Str(required=True)
+        name = fields.Str(required=True)
+        modified = fields.Str(required=True)
+
+    results = fields.List(fields.Nested(ResultsSchema), required=True)
+
+    @post_load
+    def make_object(self, data, **_kwargs):  # pylint:disable=no-self-use
+        return data["results"]
```

Comment:
This schema is doing a bunch of things for us:

1. It extracts the list of files out of the top-level `"results"` key that Blackboard sends. (The response from Blackboard is `{"results": [ <list_of_file_dicts> ]}`.)
2. It raises an exception with a correctly-formatted error message if the top-level dict has no `"results"` key or of the `"results"` key's value is not a list
3. It raises if any of the file dicts is missing any of the required fields
4. It raises if any of the file dicts has a field with an invalid value

But it should actually be doing a couple more things as well, which would simplify the service:

1. The service has to translate Blackboard API field names to the field names that our frontend expects from our proxy API: `"name"` to `"display_name"` and `"modified"` to `"updated_at"`. The schema could do this so the service doesn't have to
2. The service has to exclude any unknown fields from the results. The schema could do that
3. I'm not sure whether Blackboard's modified times are in the correct format as expected by our frontend. What format does our Canvas list files API return? The schema could translate the format. But I suspect that our frontend merely sorts them alphanumerically and doesn't actually parse the datetimes so it may be fine

Reply:
Added a [reply](https://github.com/hypothesis/lms/pull/2781#discussion_r648193154) to the thread above, but: I don't think schemas should just represent the expected data without any other functionality. They validate and _deserialize_ input objects